dbms/src/Storages/KVStore/MultiRaft/Disagg/FastAddPeer.cpp

```diff
@@ -246,136 +172,282 @@ bool tryResetPeerIdInRegion(RegionPtr region, const RegionLocalState & region_st
-                            region_id);
-                    }
+                        store_id);
                 }
             }
-            if (!success)
+        }
+        {
+            if (watch.elapsedSeconds() >= settings.fap_wait_checkpoint_timeout_seconds)
             {
-                if (watch.elapsedSeconds() >= settings.fap_wait_checkpoint_timeout_seconds)
-                {
-                    GET_METRIC(tiflash_fap_task_result, type_failed_timeout).Increment();
-                    return genFastAddPeerRes(FastAddPeerStatus::NoSuitable, "", "");
-                }
-                std::this_thread::sleep_for(std::chrono::milliseconds(1000));
+                // TODO(fap) Cancel and remove from AsyncTasks
+                // This could happen if there are too many pending tasks in queue,
+                LOG_INFO(log, "FastAddPeer timeout region_id={} new_peer_id={}", region_id, new_peer_id);
+                GET_METRIC(tiflash_fap_task_result, type_failed_timeout).Increment();
+                return genFastAddPeerRes(FastAddPeerStatus::NoSuitable, "", "");
             }
+            std::this_thread::sleep_for(std::chrono::milliseconds(1000));
         }
-        after_build();
+    }
+}
 
-        Stopwatch watch_ingest;
-        auto kvstore = server->tmt->getKVStore();
-        kvstore->handleIngestCheckpoint(region, checkpoint_info, *server->tmt);
+FastAddPeerRes FastAddPeerImplWrite(
+    TMTContext & tmt,
+    UInt64 region_id,
+    UInt64 new_peer_id,
+    CheckpointRegionInfoAndData && checkpoint,
+    UInt64 start_time)
+{
+    auto * log = &Poco::Logger::get("FastAddPeer");
+    auto fap_ctx = tmt.getContext().getSharedContextDisagg()->fap_context;
+    const auto & settings = tmt.getContext().getSettingsRef();
+
+    Stopwatch watch;
+    SCOPE_EXIT({ GET_METRIC(tiflash_fap_task_duration_seconds, type_write_stage).Observe(watch.elapsedSeconds()); });
+    GET_METRIC(tiflash_fap_task_state, type_writing_stage).Increment();
+    SCOPE_EXIT({ GET_METRIC(tiflash_fap_task_state, type_writing_stage).Decrement(); });
+
+    CheckpointInfoPtr checkpoint_info;
+    RegionPtr region;
+    RaftApplyState apply_state;
+    RegionLocalState region_state;
+    std::tie(checkpoint_info, region, apply_state, region_state) = checkpoint;
 
-        // Write raft log to uni ps
-        UniversalWriteBatch wb;
-        RaftDataReader raft_data_reader(*(checkpoint_info->temp_ps));
-        raft_data_reader.traverseRemoteRaftLogForRegion(
+    auto & storages = tmt.getStorages();
+    auto keyspace_id = region->getKeyspaceID();
+    auto table_id = region->getMappedTableID();
+    auto storage = storages.get(keyspace_id, table_id);
+    if (!storage)
+    {
+        // TODO(fap) add some ddl syncing to prevent fallback.
+        throw Exception(
+            ErrorCodes::LOGICAL_ERROR,
+            "Can't get storage engine keyspace_id={} table_id={}",
+            keyspace_id,
+            table_id);
+    }
```

Comment:
```suggestion
        RUNTIME_CHECK_MSG(storage, "Can't get storage engine for keyspace_id={} table_id={}", keyspace_id, table_id);
```

Reply:
`!storage` is expected in some cases, and I will remove the throw in later pr.